### PR TITLE
feat: add a stack for LoopBack 4 - an open source Node.js/TypeScript framework

### PR DESCRIPTION
**Pull request by raymondfeng**
See https://loopback.io

### Checklist:

- [x] Read the [Code of Conduct](https://github.com/appsody/website/blob/master/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/appsody/website/blob/master/CONTRIBUTING.md).

- [x] Followed the [commit message guidelines](https://github.com/appsody/website/blob/master/CONTRIBUTING.md#commit-message-guidelines).

- [x] Stack adheres to [Appsody stack structure](https://github.com/appsody/website/blob/master/content/docs/stacks/stack-structure.md).

### Modifying an existing stack:

- [ ] Updated the stack version in `stack.yaml`

### Contributing a new stack:

- Describe how application dependencies are managed:

> We use `npm` to manage dependencies. The image itself does not have any dependencies so far. The template includes `package.json`.

- Explain how Appsody file watcher is utilized:

> It watches `*.js` files.

- Describe other Appsody environment variables defined by the stack image:

> None.

- Describe any limitations and known issues:

> The template should be ideally allowing variables for customization but appsody does not support it.

### CODE REVIEW

File: incubator/nodejs-loopback/README.md [hunk 1]
```diff
@@ -0,0 +1,30 @@
+# Node.js LoopBack Stack
+
+The Node.js LoopBack stack extends the [Node.js stack](../nodejs/README.md) and provides a powerful solution
+to build open APIs ans micro-services in TypeScript with [LoopBack](https://loopback.io/), an open source Node.js API framework. It is based on [LoopBack 4](https://github.com/strongloop/loopback-next).
+
+## Templates
+
+- `scaffold`: The template is generated with `lb4` command from `@loopback/cli`. See https://loopback.io/doc/en/lb4/Getting-started.html for more information.
```
[the remark this answers is not in the document]
Comment (raymondfeng) [reply incubator/nodejs-loopback/README.md]: @bajtos I think you have raised a good question. I was wondering about the versioning support for stacks too. A few thoughts:

1. The quick and dirty way is to create a new stack for a new version and name it as `nodejs-loopback-v1.0.1`

2. It would be ideal to allow stacks to be published to `npm` or `maven`, which will have versioning capabilities for free. Then they can be registered or discovered.

[the remark this answers is not in the document]
Comment (neeraj-laad) [reply incubator/nodejs-loopback/README.md]: @raymondfeng @bajtos
Appsody stacks are versioned and the versioning is managed by `stack.yaml` file. The build system will use this to build new docker images and make newer temapltes available.

[the remark this answers is not in the document]
Comment (raymondfeng) [reply incubator/nodejs-loopback/README.md]: @neeraj-laad I understand the images are versioned. But can we have multiple versions of the template in the github repo? Or does it make sense?

**Comment by raymondfeng** [comment]
@neeraj-laad Thanks for the feedback and clarification. We'll look into the opportunity to divide responsibilities into stack and template. IIUC, stack provides the base image while template is a skeleton layer on top of the stack to provision a new application.

**Comment by raymondfeng** [comment]
BTW, we're working on a few extension packages to support cloud native health/metrics/tracing.

**Comment by neeraj-laad** [comment]
@raymondfeng @bajtos Please let me know if my comments are not clear. 

We are keen to have a loopback stack and https://github.com/kabanero-io/kabanero-collection/issues/6 depends on having this tack available in Appsody.

**Comment by raymondfeng** [comment]
@neeraj-laad @seabaylea I have made minimum changes so that the base image `index.js` invokes the `LB4 application` from the template per our offline discussion last Friday. 

I realized the following project layout does not work nicely with TypeScript:
```
project
  > files from base image
  - user-app
     > files from template
```

I'll have to explore further as a follow-up. For now, I just plain js in the base image.

**Comment by raymondfeng** [comment]
> We can perhaps delete the docker file from the template.

Docker related files are removed from the template. They are generated by `@loopback/cli`.

**Comment by mhamwala** [comment]
Hi @raymondfeng, I get the following error when I run the stack using `(appsody run)`.
```
[Container] Running: npm start
[Container] 
[Container] > nodejs-loopback@0.1.0 start /project
[Container] > node index.js
[Container] 
[Container] Cannot start the application. TypeError: Class constructor  cannot be invoked without 'new'
[Container]     at new DemoApplication (/project/user-app/dist/application.js:27:28)
[Container]     at main (/project/index.js:5:15)
[Container]     at Object.<anonymous> (/project/index.js:30:3)
[Container]     at Module._compile (internal/modules/cjs/loader.js:778:30)
[Container]     at Object.Module._extensions..js (internal/modules/cjs/loader.js:789:10)
[Container]     at Module.load (internal/modules/cjs/loader.js:653:32)
[Container]     at tryModuleLoad (internal/modules/cjs/loader.js:593:12)
[Container]     at Function.Module._load (internal/modules/cjs/loader.js:585:3)
[Container]     at Function.Module.runMain (internal/modules/cjs/loader.js:831:12)
[Container]     at startup (internal/bootstrap/node.js:283:19)
[Container]     at bootstrapNodeJSCore (internal/bootstrap/node.js:622:3)
[Container] npm ERR! code ELIFECYCLE
[Container] npm ERR! errno 1
[Container] npm ERR! nodejs-loopback@0.1.0 start: `node index.js`
[Container] npm ERR! Exit status 1
[Container] npm ERR! 
[Container] npm ERR! Failed at the nodejs-loopback@0.1.0 start script.
[Container] npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
[Container] 
[Container] npm ERR! A complete log of this run can be found in:
[Container] npm ERR!     /root/.npm/_logs/2019-08-14T16_25_22_536Z-debug.log
[Container] Wait received error:exit status 1
[Container] Wait received error on server start exit status 1
```

Reproduce steps:
1. `cd image`
2. `docker build -t <org-name>/<stack-id> -f Dockerfile-stack .`
3. `cd ../templates/scaffold`
4. `Edit .appsody-config.yaml to use your built stack:`
5. `appsody run`

**Comment by raymondfeng** [comment]
@neeraj-laad I cannot reproduce the problem:
```
cd image
docker build -t appsody/nodejs-loopback:0.1.0 -f Dockerfile-stack .
cd ../templates/scaffold/
appsody run
```

**Comment by mhamwala** [comment]
Okay could you try these steps on your branch:
```
1. export STACKS_LIST="incubator/nodejs-loopback"
2. cd ./ci .
3. ./build.sh ../                       (run the build script)
4. cd assets
5. `pwd` (get your current file path and add the incubator.test.yaml that has been generated on the end)
6. /Users/........./incubator-index-test.yaml
7. add `file://` at the start of that pwd: example:
appsody repo add <local-test> file:///Users/musahamwala/Documents/Code/stacks/ci/assets/incubator-index-test.yaml
8. you should be able to now init your loopback project in a new folder:
appsody init local-test/nodejs-loopback
```

**Comment by mhamwala** [comment]
let me know if these steps are unclear

**Comment by raymondfeng** [comment]
@neeraj-laad After upgrading to appsody@0.3.0, I was able to follow your instructions and get the `nodejs-loopback` stack working. I don't see the issue you experienced.

File: incubator/nodejs-loopback/templates/scaffold/package-lock.json [hunk 2]
```diff
@@ -0,0 +1,4327 @@
+{
+  "name": "nodejs-loopback-scaffold",
+  "version": "1.0.0",
+  "lockfileVersion": 1,
```
Comment (bajtos) [inline incubator/nodejs-loopback/templates/scaffold/package-lock.json]: Is it a good idea to include a package-lock file in the template?

Personally, I think developers using appsody should be getting the most recent version of all dependencies that are available at install (scaffold) time. To achieve that with a package-lock file, we would have to frequently update it (e.g. by setting up a https://renovatebot.com).

Can we simply remove the package-lock file from the template, at least for now?

File: incubator/nodejs-loopback/README.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (bajtos) [reply incubator/nodejs-loopback/README.md]: My main concern is about DX for people maintaining this template. When we introduce changes in LB4, for example we call a new sequence action from the scaffolded sequence class, or change the way how monitoring & metrics endpoints are added, how are we going to update the appsody template to match the output of the latest `lb4` CLI tool?

More specifically:
- What are the steps to determine what has changed in the output of `lb4 app` and which part of these changes to bring into appsody template?
- What process will we use to ensure the appsody template is updated after we make changes in `lb4` CLI?

I am concerned that the current proposal will make it very easy to forget to bring the appsody template up-to-date frequently enough, and also make it easy to accidentally re-introduce files like Dockerfile that we decided to remove as part of this initial pull request.

The second concern can be solved by automated test suite here in appsody that will verify that new version are not bringing back artifacts that don't belong to the template.

I would like to propose two alternative approaches:

(1)
Keep the LB template inside loopback-next monorepo, so that we can make changes both to `lb4` CLI and appsody template as part of a single pull request. Add a new item to our PR checklist to remind us to update the appsody template (see https://raw.githubusercontent.com/strongloop/loopback-next/master/.github/PULL_REQUEST_TEMPLATE.md)

(2)
Don't lock the LB template inside appsody, modify the project generator to call `lb4` CLI to create project files. We can introduce a new flag to `lb4 app` command to tell it to create appsody-based project.

File: incubator/nodejs-loopback/templates/scaffold/package-lock.json (hunk 2 above)
Comment (neeraj-laad) [reply incubator/nodejs-loopback/templates/scaffold/package-lock.json]: I agree with this comment. Even though there is a likelihood that latest version might be the one stack provider has tested on, it is preferred that they get the latest versions within the rules allowed by package.json.

**Comment by neeraj-laad** [comment]
@bajtos how is the version oc loopback cli controller in this stack/template today?

I'm wondering if we can use Appsody init hook to create the loopback project  when the user does an `appsody init` vs checking in a specific version. 

That said we've tried to avoid lot of generated code that user's have to get rid of as a general principle for other stacks, so if there is a way without it that would be preferred. 

Note: I'm not a loopback user, so feel free to say if this does not make sense for loopback framework.

**Comment by neeraj-laad** [comment]
@raymondfeng I can build the stack and templates fine and run the app with those.

**Comment by raymondfeng** [comment]
I had some offline discussion with @neeraj-laad and @ebullient. Here is a brief summary:

1. It's possible to host LoopBack images/templates outside of `appsody/stacks` repo. The registration can be done via a PR to update https://github.com/appsody/stacks/blob/master/index.yaml. The `url` should point to a stack release. 

@ebullient pointed out:
> So. it is possible .. but it breaks down some of the automated tooling in ./ci/package.sh and ./ci/release.sh 

2. @ebullient proposed that the `base` image can be maintained in a separate repo and published to docker hub as part of the release process. In `appsody/stacks/incubator/nodejs-loopback`, we keep the image, which will be a very light shim to the base image from LoopBack. That makes `incubator/nodejs-loopback/image/Dockerfile-stack` very simple with `FROM`.

**Comment by raymondfeng** [comment]
> @neeraj-laad I'm wondering if we can use Appsody init hook to create the loopback project when the user does an appsody init vs checking in a specific version.

How does `init hook` work? What does it take to contribute/register a hook?

**Comment by mhamwala** [comment]
we can add a `launch.json` file inside the .vscode folder for attaching the debugger in vscode
```
{
  // Use IntelliSense to learn about possible attributes.
  // Hover to view descriptions of existing attributes.
  // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
  "version": "0.2.0",
  "configurations": [{
    "name": "Appsody: Attach node debugger",
    "type": "node",
    "request": "attach",
    "port": 9229,
    "remoteRoot": "/project/user-app",
    "localRoot": "${workspaceFolder}",
    "protocol": "inspector"
  }]
}
```

File: incubator/nodejs-loopback/README.md [hunk 3]
```diff
@@ -0,0 +1,62 @@
+# Node.js LoopBack Stack
+
+The Node.js LoopBack stack extends the [Node.js stack](../nodejs/README.md) and provides a powerful solution
```
Comment (mhamwala) [inline incubator/nodejs-loopback/README.md]: nit: point this link to the `../nodejs`

File: incubator/nodejs-loopback/README.md [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+   You can continue to edit the application in your preferred IDE (VSCode or others) and your changes will be reflected in the running container within a few seconds.
+
+
+## Test the stack locally
+
+```sh
+cd nodejs-loopback
+cd image
+docker build -t appsody/nodejs-loopback:0.1.0 -f Dockerfile-stack .
```
Comment (mhamwala) [inline incubator/nodejs-loopback/README.md]: I think we should either change this to (as the stack version is bound to change):
```
appsody/nodejs-loopback:0.1.0 -> <org-name>/<stack-id>
Edit .appsody-config.yaml to use your built stack:
stack: <org-name>/<stack-id>
```
or link to: https://appsody.dev/docs/stacks/modify

File: incubator/nodejs-loopback/image/Dockerfile-stack [hunk 5]
```diff
@@ -0,0 +1,34 @@
+FROM appsody/nodejs:0.2
+
+ENV APPSODY_MOUNTS=/:/project/user-app
+ENV APPSODY_DEPS=/project/user-app/node_modules
+
+ENV APPSODY_WATCH_DIR=/project/user-app
+ENV APPSODY_WATCH_IGNORE_DIR=/project/user-app/node_modules
+ENV APPSODY_WATCH_REGEX="^.*.js$"
```
Comment (mhamwala) [inline incubator/nodejs-loopback/image/Dockerfile-stack]: should we be watching `.ts` files here? my understanding is that we need to watch .ts files, otherwise, it will not automatically rebuild your TS? and the .js files would stay the same but I could be wrong?

File: incubator/nodejs-loopback/stack.yaml [hunk 6]
```diff
@@ -0,0 +1,7 @@
+id: nodejs-loopback
+name: LoopBack 4 application
+version: 0.1.0
+description: LoopBack 4 application
```
Comment (mhamwala) [inline incubator/nodejs-loopback/stack.yaml]: description suggestion (this will appear on the website)
> Loopback 4 API framework for Node.js

File: incubator/nodejs-loopback/stack.yaml [hunk 7]
```diff
@@ -0,0 +1,7 @@
...
+maintainer: Raymond Feng <enjoyjava@gmail.com>
```
Comment (mhamwala) [inline incubator/nodejs-loopback/stack.yaml]: Could you change this to the following format:
```
maintainers:
  - name: Raymond Feng
    email: enjoyjava@gmail.com
    github-id: raymondfeng
```

File: incubator/nodejs-loopback/templates/scaffold/.appsody-config.yaml [hunk 8]
```diff
@@ -0,0 +1 @@
+stack: appsody/nodejs-loopback:0.1.0
```
Comment (mhamwala) [inline incubator/nodejs-loopback/templates/scaffold/.appsody-config.yaml]: could you remove the `.0` on the end so that it always picks up the latest:
eg: 
> stack: appsody/nodejs-loopback:0.1

File: incubator/nodejs-loopback/stack.yaml [hunk 9]
```diff
@@ -0,0 +1,7 @@
+id: nodejs-loopback
+name: LoopBack 4 application
```
Comment (mhamwala) [inline incubator/nodejs-loopback/stack.yaml]: This name will be used on the website, possible suggestion:
> Node.js LoopBack 4

File: incubator/nodejs-loopback/templates/scaffold/.yo-rc.json [hunk 10]
```diff
@@ -0,0 +1 @@
+{}
```
Comment (mhamwala) [inline incubator/nodejs-loopback/templates/scaffold/.yo-rc.json]: this is file needed?

File: incubator/nodejs-loopback/templates/scaffold/README.md [hunk 11]
```diff
@@ -0,0 +1,3 @@
+# nodejs-loopback-scaffold
```
Comment (mhamwala) [inline incubator/nodejs-loopback/templates/scaffold/README.md]: is this file needed?

File: incubator/nodejs-loopback/templates/scaffold/src/__tests__/README.md [hunk 12]
```diff
@@ -0,0 +1,3 @@
+# Tests
```
Comment (mhamwala) [inline incubator/nodejs-loopback/templates/scaffold/src/__tests__/README.md]: we could possibly remove this readme?

File: incubator/nodejs-loopback/templates/scaffold/src/models/README.md [hunk 13]
```diff
@@ -0,0 +1,3 @@
+# Models
```
Comment (mhamwala) [inline incubator/nodejs-loopback/templates/scaffold/src/models/README.md]: perhaps we can remove this readme?

File: incubator/nodejs-loopback/templates/scaffold/src/repositories/README.md [hunk 14]
```diff
@@ -0,0 +1,3 @@
+# Repositories
```
Comment (mhamwala) [inline incubator/nodejs-loopback/templates/scaffold/src/repositories/README.md]: same here

File: incubator/nodejs-loopback/templates/scaffold/src/datasources/README.md [hunk 15]
```diff
@@ -0,0 +1,3 @@
+# Datasources
```
Comment (mhamwala) [inline incubator/nodejs-loopback/templates/scaffold/src/datasources/README.md]: we could remove this readme?

File: incubator/nodejs-loopback/templates/scaffold/src/controllers/README.md [hunk 16]
```diff
@@ -0,0 +1,9 @@
+# Controllers
```
Comment (mhamwala) [inline incubator/nodejs-loopback/templates/scaffold/src/controllers/README.md]: we could remove this readme?

File: incubator/nodejs-loopback/templates/scaffold/.yo-rc.json (hunk 10 above)
Comment (raymondfeng) [reply incubator/nodejs-loopback/templates/scaffold/.yo-rc.json]: Updated. It tracks the `@loopback/cli` version.

File: incubator/nodejs-loopback/templates/scaffold/src/__tests__/README.md (hunk 12 above)
Comment (raymondfeng) [reply incubator/nodejs-loopback/templates/scaffold/src/__tests__/README.md]: We generate these files to guide developers to understand purposes of such folders. I would like to keep them to be consistent with `@loopback/cli`.

**Comment by raymondfeng** [comment]
@mhamwala Thank you for the feedback. PTAL.

**Comment by mhamwala** [comment]
could you add this to the loopback `stack.yaml` (reason its failing the ci script):
> language: nodejs

File: incubator/nodejs-loopback/stack.yaml [hunk 17]
```diff
@@ -1,7 +1,10 @@
 id: nodejs-loopback
-name: LoopBack 4 application
+name: LoopBack 4 Framework for Node.js
```
Comment (mhamwala) [inline incubator/nodejs-loopback/stack.yaml]: nit: LoopBack 4 Node.js Framework?

I'm for keeping the name as short as possible. take a look at the current stacks on https://appsody.dev/

File: incubator/nodejs-loopback/stack.yaml [hunk 18]
```diff
@@ -1,7 +1,10 @@
...
 version: 0.1.0
-description: LoopBack 4 application
-maintainer: Raymond Feng <enjoyjava@gmail.com>
+description: LoopBack 4 Framework for Node.js - loopback.io
```
Comment (mhamwala) [inline incubator/nodejs-loopback/stack.yaml]: nit: looks too similar to the title
> LoopBack 4 API Framework for Node.js

**Comment by neeraj-laad** [comment]
@raymondfeng Can we move this stack to the incubator repo given it is fairly new and needs a bit more work.

**Comment by raymondfeng** [comment]
@neeraj-laad The stack is already under `incubator`, isn't it?

**Comment by neeraj-laad** [comment]
Sorry I mean moving it to experimental repo

**Comment by raymondfeng** [comment]
@neeraj-laad I moved it to `experimental` and cleaned up the commit history. PTAL.
